## [作業重點]
確保你了解隨機森林模型中每個超參數的意義，並觀察調整超參數對結果的影響

## 作業

1. 試著調整 RandomForestClassifier(...) 中的參數，並觀察是否會改變結果？
2. 改用其他資料集 (boston, wine)，並與回歸模型與決策樹的結果進行比較

In [1]:
from sklearn import datasets, metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

In [2]:
# 讀取鳶尾花資料集
iris = datasets.load_iris()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=6)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [3]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9736842105263158


In [4]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [5]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.11122828 0.03797093 0.33459306 0.51620773]


作業¶
試著調整 RandomForestClassifier(...) 中的參數，並觀察是否會改變結果？
改用其他資料集 (boston, wine)，並與回歸模型與決策樹的結果進行比較

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=4)
使用 20 顆樹Accuracy:  0.9473684210526315 , 使用 30 顆樹0.9736842105263158
    
每棵樹的最大深度為 4Accuracy:  0.9473684210526315 , 每棵樹的最大深度為 6        0.9736842105263158

In [7]:

from sklearn.ensemble import RandomForestRegressor
# 讀取資料集
boston = datasets.load_boston()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(boston.data, boston.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestRegressor(n_estimators=20, max_depth=4)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [8]:
print(clf.score(x_train,y_train))
print(clf.score(x_test,y_test))

0.9077135975085595
0.8127892831919981


使用回歸模型與決策樹 score  0.9736842105263158
使用隨機森林模 score  0.8127892831919981  
使用隨機森林模型效果較佳

In [9]:
# 讀取資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=4)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [10]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9777777777777777


In [11]:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [12]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.17023715 0.02856356 0.01211127 0.02131765 0.01401176 0.0148574
 0.15868108 0.02371127 0.01922802 0.1487247  0.1033471  0.09695948
 0.18824956]


使用回歸模型與決策樹 Accuracy:  0.9736842105263158
使用隨機森林模 Accuracy:  0.9777777777777777   
使用隨機森林模型效果較佳